backend/src/packages/chaiNNer_standard/image_filter/noise/fast_nlmeans.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+import numpy as np
+import cv2
+
+from nodes.properties.inputs import ImageInput, SliderInput, NumberInput
+from nodes.properties.outputs import ImageOutput
+from nodes.groups import Condition, if_group
+from nodes.impl.image_utils import to_uint8
+from nodes.utils.utils import get_h_w_c
+
+from .. import noise_group
+
+
+@noise_group.register(
+    schema_id="chainner:image:fast_nlmeans",
+    name="Fast NL means",
+    description="Use the fast Non-Local Means algorithm to denoise an image.",
+    icon="CgEditNoise",
+    inputs=[
+        ImageInput("Image", channels=[1, 3, 4]),
+        SliderInput(
+            "Luminance strength",
+            minimum=0.,
+            maximum=50.,
+            default=3.0,
+            precision=1,
+            controls_step=0.1,
+            slider_step=0.1,
+        ),
+        if_group(
+            Condition.type(0, "Image { channels: 3 }")
+            | Condition.type(0, "Image { channels: 4 }")
+        )(
+            SliderInput(
+                "Color strength",
+                minimum=0.,
+                maximum=50.,
+                default=3.,
+                precision=1,
+                controls_step=0.1,
+                slider_step=0.1,
+            )
+        ),
+        NumberInput("Template window size",
+            minimum=1,
+            default=7,
+            maximum=49,
+            precision=0,
+            controls_step=2),
+        NumberInput("Search window size",
+            minimum=1,
+            default=21,
+            maximum=49,
+            precision=0,
+            controls_step=2),
+    ],
+    outputs=[ImageOutput(image_type="Input0")],
+)
+def fast_nlmeans_node(
+    img: np.ndarray,
+    h: float,
+    h_color: float,
+    template_window_size: int,
+    search_window_size: int,
+) -> np.ndarray:
+
+    if not (template_window_size % 2):
+        raise ValueError("\'Template window size\' must be an odd value.")
+    if not (search_window_size % 2):
+        raise ValueError("\'Search window size\' must be an odd value.")
```

Comment:
Instead of asserting odd values here, I think we should instead think of a way to allow all inputs. The main advantage of all allowing all values in the input range is that the user experience is better. The type system cannot enforce odd values, so users cannot know that e.g. 20 is not valid until they run their chain and see it crash.

One idea would to use a radius instead of a size. Then we can then do 2r+1 to get an odd size.